File: scripts/solar/util/math/InterpolationUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 6,["9"] = 6,["10"] = 6,["12"] = 6,["13"] = 15,["14"] = 16,["15"] = 15,["16"] = 21,["17"] = 22,["18"] = 23,["19"] = 23,["20"] = 23,["21"] = 23,["22"] = 21,["23"] = 32,["24"] = 32,["25"] = 33,["27"] = 34,["28"] = 34,["29"] = 35,["30"] = 34,["33"] = 37,["34"] = 38,["36"] = 40,["37"] = 40,["38"] = 40,["39"] = 40,["41"] = 32,["42"] = 6});
local ____exports = {}
local ____VectorUtil = require("solar.util.math.VectorUtil")
local VectorUtil = ____VectorUtil.default
____exports.default = __TS__Class()
local InterpolationUtil = ____exports.default
InterpolationUtil.name = "InterpolationUtil"
function InterpolationUtil.prototype.____constructor(self)
end
function InterpolationUtil.line(self, start, ____end, percent)
    return start + (____end - start) * percent
end
function InterpolationUtil.vecIn2Vectors(self, start, ____end, percent)
    local vec = VectorUtil:subtract(____end, start)
    return VectorUtil:add(
        start,
        VectorUtil:multScalar(vec, percent)
    )
end
function InterpolationUtil.bezier(self, percent, ...)
    local controlPoints = {...}
    local tempVS = {}
    do
        local i = 1
        while i < #controlPoints do
            tempVS[i] = ____exports.default:vecIn2Vectors(controlPoints[i], controlPoints[i + 1], percent)
            i = i + 1
        end
    end
    if #tempVS == 1 then
        return tempVS[1]
    else
        return ____exports.default:bezier(
            percent,
            table.unpack(tempVS)
        )
    end
end
____exports.default = InterpolationUtil
return ____exports

File: src/solar/util/math/InterpolationUtil.ts
import VectorUtil from "./VectorUtil";

/**
 * 插值工具
 */
export default class InterpolationUtil {


    /**
     * 线性插值
     * @param start
     * @param end
     * @param percent
     */
    static line(start: number, end: number, percent: number): number {
        return start + ((end - start) * percent);
    }


    //两点间 percent 比例的点
    static vecIn2Vectors(start: Vector, end: Vector, percent: number): Vector {
        let vec = VectorUtil.subtract(end, start)
        return VectorUtil.add(start, VectorUtil.multScalar(vec, percent))
    }


    /**
     * 贝塞尔插值
     * @param percent
     * @param controlPoints
     */
    static bezier(percent: number, ...controlPoints: Vector[]) {
        let tempVS: Vector[] = []
        for (let i = 1; i < controlPoints.length; i++) {
            tempVS[i - 1] = InterpolationUtil.vecIn2Vectors(controlPoints[i - 1], controlPoints[i], percent)
        }
        if (tempVS.length == 1) {
            return tempVS[0]
        } else {
            return InterpolationUtil.bezier(percent, ...tempVS)
        }
    }
}